File: swift/BCComponents/Sources/BCComponents/KeyDerivationMethod.swift
import DCBOR

public enum KeyDerivationMethod: UInt64, CaseIterable, Equatable, Hashable, Sendable {
    case hkdf = 0
    case pbkdf2 = 1
    case scrypt = 2
    case argon2id = 3

    public static var `default`: KeyDerivationMethod {
        .argon2id
    }

    public func index() -> Int {
        Int(rawValue)
    }

    public static func fromIndex(_ index: Int) -> KeyDerivationMethod? {
        KeyDerivationMethod(rawValue: UInt64(index))
    }
}

extension KeyDerivationMethod: CustomStringConvertible {
    public var description: String {
        switch self {
        case .hkdf:
            return "HKDF"
        case .pbkdf2:
            return "PBKDF2"
        case .scrypt:
            return "Scrypt"
        case .argon2id:
            return "Argon2id"
        }
    }
}

extension KeyDerivationMethod: CBOREncodable {
    public var cbor: CBOR {
        .unsigned(rawValue)
    }
}

extension KeyDerivationMethod: CBORDecodable {
    public init(cbor: CBOR) throws {
        let index = try Int(cbor: cbor)
        guard let method = KeyDerivationMethod.fromIndex(index) else {
            throw BCComponentsError.general("Invalid KeyDerivationMethod")
        }
        self = method
    }
}

File: typescript/bc-components/src/encrypted-key/key-derivation-method.ts
import { type Cbor } from '@bc/dcbor';

import { BCComponentsError } from '../error.js';

export enum KeyDerivationMethod {
    HKDF = 0,
    PBKDF2 = 1,
    Scrypt = 2,
    Argon2id = 3,
}

export function keyDerivationMethodIndex(method: KeyDerivationMethod): number {
    return method as number;
}

export function keyDerivationMethodFromIndex(index: number): KeyDerivationMethod | undefined {
    switch (index) {
        case 0:
            return KeyDerivationMethod.HKDF;
        case 1:
            return KeyDerivationMethod.PBKDF2;
        case 2:
            return KeyDerivationMethod.Scrypt;
        case 3:
            return KeyDerivationMethod.Argon2id;
        default:
            return undefined;
    }
}

export function keyDerivationMethodFromCbor(cbor: Cbor): KeyDerivationMethod {
    const index = Number(cbor.toInteger());
    const method = keyDerivationMethodFromIndex(index);
    if (method === undefined) {
        throw BCComponentsError.general('Invalid KeyDerivationMethod');
    }
    return method;
}

export function keyDerivationMethodToString(method: KeyDerivationMethod): string {
    switch (method) {
        case KeyDerivationMethod.HKDF:
            return 'HKDF';
        case KeyDerivationMethod.PBKDF2:
            return 'PBKDF2';
        case KeyDerivationMethod.Scrypt:
            return 'Scrypt';
        case KeyDerivationMethod.Argon2id:
            return 'Argon2id';
    }
}
